File: P-3/python/p3.py
import unittest

def swap_beginning_end(l):
    print("sport", end="")
    beginning = l[:int(len(l)/2)]
    end = l[int(len(l)/2):]
    return end + beginning

def value_to_index(l):
    print("hawk", end=" ")
    return [l[i] for i in l]

def swap_first_last(l):
    print("glass", end=" ")
    first = l.pop(0)
    last = l.pop()
    return [last] + l + [first]

def do_si_do(l):
    print("snap", end=" ")
    mixed = []
    it = iter(l)
    for e in it:
        nxt = next(it, None)
        if nxt is not None:
            mixed.append(nxt)
        mixed.append(e)
    return mixed

def rotate(l):
    print("gossip", end=" ")
    last = l.pop()
    return [last] + l

def shuffle(l):
    print("alter", end=" ")
    mixed = []
    beginning = l[:int(len(l)/2)]
    end = l[int(len(l)/2):]
    for (i, e) in enumerate(end):
        mixed.append(e)
        if i < len(beginning):
            mixed.append(beginning[i])
    return mixed

def substitution_table(l):
    print("avocado", end=" ")
    return [l[4], l[0], l[2], l[3], l[1], l[5]]

def substitution_table2(l):
    print("alcohol", end=" ")
    return [l[5], l[1], l[2], l[0], l[4], l[3]]

def twin_propeller(l):
    print("dolphin", end=" ")
    beginning = l[:int(len(l)/2)]
    end = l[int(len(l)/2):]
    beginning.reverse()
    end.reverse()
    return beginning + end

def fold_shuffle(l):
    print("material", end=" ")
    mixed = []
    beginning = l[:int(len(l)/2)]
    end = l[int(len(l)/2):]
    beginning.reverse()
    end.reverse()
    for (i, e) in enumerate(end):
        mixed.append(e)
        if i < len(beginning):
            mixed.append(beginning[i])
    return mixed

def propeller_shuffle(l):
    print("scene", end=" ")
    mixed = []
    beginning = l[:int(len(l)/2)]
    beginning.reverse()
    end = l[int(len(l)/2):]
    for (i, e) in enumerate(end):
        mixed.append(e)
        if i < len(beginning):
            mixed.append(beginning[i])
    return mixed

def reverse(l):
    print("floor", end=" ")
    l.reverse()
    return l


def run_first_6(l):
    l = xxxxxxxx(l)
    l = xxxxxxxx(l)
    l = xxxxxxxx(l)
    l = xxxxxxxx(l)
    l = xxxxxxxx(l)
    l = xxxxxxxx(l)
    return l

def run_second_6(l):
    l = xxxxxxxx(l)
    l = xxxxxxxx(l)
    l = xxxxxxxx(l)
    l = xxxxxxxx(l)
    l = xxxxxxxx(l)
    l = xxxxxxxx(l)
    return l

class Tests(unittest.TestCase):
    def test_a(self):
        self.assertEqual(run_first_6([0, 1, 2, 3, 4, 5]), [3, 0, 1, 4, 5, 2])

    # clw answerwallet --passphrase "X Y Z"
    # address 131ykwoBx3uPiWwUEWw63KYLLjX93d3L1E
    def test_b(self):
        self.assertEqual(run_second_6(run_first_6([0, 1, 2, 3, 4, 5])), [5, 4, 3, 2, 1, 0])

unittest.main()

File: P-3/javascript/p3.js
const swap_beginning_end = (l) => {
  process.stdout.write('sport ')
  const beginning = l.slice(0, l.length / 2)
  const end = l.slice(l.length / 2, l.length)
  return end.concat(beginning)
}

const value_to_index = (l) => {
  process.stdout.write('hawk ')
  return l.map(e => l[e])
}

const swap_first_last = (l) => {
  process.stdout.write('glass ')
  const first = l.shift()
  const last = l.pop()
  return [last].concat(l).concat(first)
}

const do_si_do = (l) => {
  process.stdout.write('snap ')
  const mixed = []
  for (let i = 0; i < l.length; i += 2) {
    const e = l[i]
    if (i + 1 < l.length) {
      const nxt = l[i + 1]
      mixed.push(nxt)
    }
    mixed.push(e)
  }
  return mixed
}

const rotate = (l) => {
  process.stdout.write('gossip ')
  l.unshift(l.pop())
  return l
}

const shuffle = (l) => {
  process.stdout.write('alter ')
  const mixed = []
  const beginning = l.slice(0, l.length / 2)
  const end = l.slice(l.length / 2, l.length)
  for (let i = 0; i < end.length; i++) {
    const e = end[i]
    mixed.push(e)
    if (i < beginning.length) {
      mixed.push(beginning[i])
    }
  }
  return mixed
}

const substitution_table = (l) => {
  process.stdout.write('avocado ')
  return [l[4], l[0], l[2], l[3], l[1], l[5]]
}

const substitution_table2 = (l) => {
  process.stdout.write('alcohol ')
  return [l[5], l[1], l[2], l[0], l[4], l[3]]
}

const twin_propeller = (l) => {
  process.stdout.write('dolphin ')
  const beginning = l.slice(0, l.length / 2)
  const end = l.slice(l.length / 2, l.length)
  beginning.reverse()
  end.reverse()
  return beginning.concat(end)
}

const fold_shuffle = (l) => {
  process.stdout.write('material ')
  const mixed = []
  const beginning = l.slice(0, l.length / 2)
  const end = l.slice(l.length / 2, l.length)
  beginning.reverse()
  end.reverse()
  for (let i = 0; i < end.length; i++) {
    const e = end[i]
    mixed.push(e)
    if (i < beginning.length) {
      mixed.push(beginning[i])
    }
  }
  return mixed
}

const propeller_shuffle = (l) => {
  process.stdout.write('scene ')
  const mixed = []
  const beginning = l.slice(0, l.length / 2)
  beginning.reverse()
  const end = l.slice(l.length / 2, l.length)
  for (let i = 0; i < end.length; i++) {
    const e = end[i]
    mixed.push(e)
    if (i < beginning.length) {
      mixed.push(beginning[i])
    }
  }
  return mixed
}

const reverse = (l) => {
  process.stdout.write('floor ')
  l.reverse()
  return l
}

const run_first_6 = (l) => {
  l = xxxxxxxx(l)
  l = xxxxxxxx(l)
  l = xxxxxxxx(l)
  l = xxxxxxxx(l)
  l = xxxxxxxx(l)
  l = xxxxxxxx(l)
  return l
}

const run_second_6 = (l) => {
  l = xxxxxxxx(l)
  l = xxxxxxxx(l)
  l = xxxxxxxx(l)
  l = xxxxxxxx(l)
  l = xxxxxxxx(l)
  l = xxxxxxxx(l)
  return l
}

const test_A = () => {
  return run_first_6([0, 1, 2, 3, 4, 5]).toString() === [3, 0, 1, 4, 5, 2].toString()
}

// clw answerwallet --passphrase "X Y Z"
// address 131ykwoBx3uPiWwUEWw63KYLLjX93d3L1E
const test_B = () => {
  return run_second_6(run_first_6([0, 1, 2, 3, 4, 5])).toString() === [5, 4, 3, 2, 1, 0].toString()
}

if (test_A() && test_B()) {
  console.log('PASS')
} else {
  console.log('FAIL')
}
